updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Result;
+use async_ctrlc::CtrlC;
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use std::{collections::HashMap, net::Shutdown, process};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "updatehub-statechange.sock";
+
+type StateChangeCallback = fn(State);
+type ErrorCallback = fn(String);
```

Comment:
We should use the [fn type](https://doc.rust-lang.org/stable/std/primitive.fn.html), we should use the [Fn trait](https://doc.rust-lang.org/stable/std/ops/trait.Fn.html) as the traits can be more flexible to allow things like closures that capture environment or custom types that implement `Fn` to be "callable".